bindings/java/src/integration/com/apple/foundationdb/CycleMultiClientIntegrationTest.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * CycleMultiClientIntegrationTest
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.apple.foundationdb;
+
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import com.apple.foundationdb.tuple.Tuple;
+
+import org.junit.jupiter.api.Assertions;
+
+/**
+ * Setup: Generating a cycle 0 -> 1 -> 2 -> 3 -> 0, its length is 4
+ * Process: randomly choose an element, reverse 2nd and 4rd element, considering the chosen one as the 1st element.
+ * Check: verify no element is lost or added, and they are still a cycle.
+ * 
+ * This test is to verify the atomicity of transactions. 
+ */
+public class CycleMultiClientIntegrationTest {
+    public static final MultiClientHelper clientHelper = new MultiClientHelper();
+
+    // more write txn than validate txn, as parent thread waits only for validate txn.
+    private static final int writeTxnCnt = 2000;
+    private static final int validateTxnCnt = 1000;
+    private static final int threadPerDB = 5;
+
+    private static final int cycleLength = 4;
+    private static List<String> expected = new ArrayList<>(Arrays.asList("0", "1", "2", "3"));
+
+    public static void main(String[] args) throws Exception {
+        FDB fdb = FDB.selectAPIVersion(630);
+        setupThreads(fdb);
+        Collection<Database> dbs = clientHelper.openDatabases(fdb); // the clientHelper will close the databases for us
+        System.out.println("Starting tests");
+        setup(dbs);
+        System.out.println("Start processing and validating");
+        process(dbs);
+        check(dbs);
```

Comment:
Can you run process + check parallel instead?  Otherwise, this looks good.

Reply:
I also verified by printing some logs

```
3 check  db com.apple.foundationdb.FDBDatabase@16d76970
3 check  db com.apple.foundationdb.FDBDatabase@16d76970
2 check  db com.apple.foundationdb.FDBDatabase@4c5a9991
2 check  db com.apple.foundationdb.FDBDatabase@5859f3cd
4 check  db com.apple.foundationdb.FDBDatabase@16d76970
2 check  db com.apple.foundationdb.FDBDatabase@4c5a9991
4 check  db com.apple.foundationdb.FDBDatabase@16d76970
4 check  db com.apple.foundationdb.FDBDatabase@16d76970
4 check  db com.apple.foundationdb.FDBDatabase@16d76970
4 check  db com.apple.foundationdb.FDBDatabase@16d76970
2 workload   db com.apple.foundationdb.FDBDatabase@5859f3cd
3 check  db com.apple.foundationdb.FDBDatabase@5859f3cd
3 check  db com.apple.foundationdb.FDBDatabase@5859f3cd
3 check  db com.apple.foundationdb.FDBDatabase@5859f3cd
5 check  db com.apple.foundationdb.FDBDatabase@16d76970
2 workload   db com.apple.foundationdb.FDBDatabase@16d76970
3 check  db com.apple.foundationdb.FDBDatabase@5859f3cd
2 workload   db com.apple.foundationdb.FDBDatabase@4c5a9991
```

Reply:
`workload` is when `process` thread runs, `check` is when `check` thread runs, it does seem that we have a little too many checks though.